backend/src/apiserver/server/run_server_test.go

```diff
@@ -455,6 +456,19 @@ func TestValidateCreateRunRequest_EmptyName(t *testing.T) {
 	assert.Contains(t, err.Error(), "The run name is empty")
 }
 
+func TestValidateCreateRunRequest_InvalidPipelineVersionReference(t *testing.T) {
```

Comment:
Not sure whether I need to duplicate the UT for job and run since the UT of such logic have been covered in UT of validatePipelineSpecAndCheckPipelineVersionReference

Reply:
This is a good question. I don't have a good answer either.
I usually test all the corner cases in the lowest level UT (e.g. UT of validatePipeline) and only verify e2e integration is correct in UTs that integrate more components (e.g. UT of job/run server).

Reply:
My main concern is that usually we don't test private methods(FYI: go/unit-test-practices#public-apis).
Ideally ,we should make validatePipeline() private since it's only used in the package. When we make it private, usually we won't write UT for it. We will write UT for public methods like CreateRunRequest() and CreateJob(). Therefore, we have to duplicate the UT logic for CreateRunRequest() and CreateJob().

Reply:
By the way, I think it may be better if we put https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/util.go under its own package like util and make some util methods public.

Reply:
discussed offline. Will move https://github.com/kubeflow/pipelines/blob/master/backend/src/apiserver/server/util.go under its own package like util and make some util methods public in another PR.